pkg/proxy/conntrack/cleanup.go

```diff
@@ -20,95 +20,109 @@ limitations under the License.
+			// not take traffic policies or topology of the service into account.
+			// This ensures that the behavior of UDP services remains consistent
+			// with TCP services.
+			if endpoint.IsServing() {
+				endpointIPs.Insert(endpoint.IP())
+			}
+		}
+
+		// populate the indices
+		ipIndex[svc.ClusterIP().String()] = endpointIPs
+		for _, loadBalancerIP := range svc.LoadBalancerVIPs() {
+			ipIndex[loadBalancerIP.String()] = endpointIPs
+		}
+		for _, externalIP := range svc.ExternalIPs() {
+			ipIndex[externalIP.String()] = endpointIPs
+		}
+		if svc.NodePort() != 0 {
+			portIndex[svc.NodePort()] = endpointIPs
+		}
 	}
-}
 
-// deleteStaleEndpointConntrackEntries takes care of flushing stale conntrack entries related
-// to UDP endpoints. After a UDP endpoint is removed we must flush any conntrack entries
-// for it so that if the same client keeps sending, the packets will get routed to a new endpoint.
-func deleteStaleEndpointConntrackEntries(ct Interface, ipFamily v1.IPFamily, svcPortMap proxy.ServicePortMap, endpointsUpdateResult proxy.UpdateEndpointsMapResult) {
 	var filters []netlink.CustomConntrackFilter
-	for _, epSvcPair := range endpointsUpdateResult.DeletedUDPEndpoints {
-		if svcInfo, ok := svcPortMap[epSvcPair.ServicePortName]; ok {
-			endpointIP := proxyutil.IPPart(epSvcPair.Endpoint)
-			nodePort := svcInfo.NodePort()
-			if nodePort != 0 {
-				filters = append(filters, filterForPortNAT(endpointIP, nodePort, v1.ProtocolUDP))
+	for _, entry := range entries {
+		// we only deal with UDP protocol entries
+		if entry.Forward.Protocol != protocolMap[v1.ProtocolUDP] {
+			continue
+		}
 
+		origDst := entry.Forward.DstIP.String()
+		origPortDst := int(entry.Forward.DstPort)
+		replySrc := entry.Reverse.SrcIP.String()
+
+		// if the original destination (--orig-dst) matches one of ClusterIP, LoadBalancerIPs and ExternalIPs
+		// of the service and the DNATed destination (--reply-src) is not in the allowed set, we clear the
+		// conntrack flow entry.
+		if _, ok := ipIndex[origDst]; ok {
+			if !ipIndex[origDst].Has(replySrc) {
+				filters = append(filters, filterForNAT(origDst, replySrc, v1.ProtocolUDP))
 			}
-			filters = append(filters, filterForNAT(svcInfo.ClusterIP().String(), endpointIP, v1.ProtocolUDP))
-			for _, extIP := range svcInfo.ExternalIPs() {
-				filters = append(filters, filterForNAT(extIP.String(), endpointIP, v1.ProtocolUDP))
-			}
-			for _, lbIP := range svcInfo.LoadBalancerVIPs() {
-				filters = append(filters, filterForNAT(lbIP.String(), endpointIP, v1.ProtocolUDP))
+		}
+
+		// if the original port destination (--orig-port-dst) matches NodePort of the service
+		// and the DNATed destination (--reply-src) is not in the allowed set, we clear the
+		// conntrack flow entry.
+		if _, ok := portIndex[origPortDst]; ok {
+			if !portIndex[origPortDst].Has(replySrc) {
+				filters = append(filters, filterForPortNAT(replySrc, origPortDst, v1.ProtocolUDP))
 			}
 		}
 	}
```

Comment:
This is deleting NAT entries for endpoints that have been deleted, but it doesn't handle deleting black-hole entries for endpoints that exist. (The stuff that was using `filterforIP` and `filterForPort` before, which you delete below.)

If a service has at least one endpointIP, then there should not be any black-hole entries for that service's clusterIP, externalIPs, LB IPs, or NodePorts.

Reply:
Whenever any endpoint of the service is deleted, after networking programming the DNAT flows of that endpoint will be cleaned by the reconciler. Doesn't this guarantee that there won't be any blackhole entries?

> but it doesn't handle deleting black-hole entries for endpoints that exist

There can't be any black-hole entries for the endpoints that exists (serving state), right? A serving endpoint will always receive packets.

Reply:
My thought process:
If a service has no endpoints then endpointsMap will be empty, after programming the rules reconciler will clear all the conntrack flow for that serviceIP. The next packet to the ServiceIP won't be DNATed as there won't be any DNAT rule (no conntrack flow) so we don't need `filterforIP` and `filterForPort`. 

If an endpoint exists and is serving then that flow is non-blackholing, assuming that service change should affect only new connections not existing once.

Reply:
If the service is deleted we leave the existing flows, just like TCP so we intentionally don't delete the flow making it consistent with the TCP flow. 
https://github.com/kubernetes/kubernetes/issues/126130#issuecomment-2332493443

Reply:
Ah, this explains why the comments were wrong too. You're misunderstanding some of the details.

From kube-proxy's point of view, there are _two_ different kinds of conntrack entries it needs to deal with:
- "DNAT entries" that get created when iptables/nftables DNATs a packet, which record things like "packets from client 10.0.1.8:12345 to service 172.30.0.10:53 get NATted to endpoint 10.128.0.5:5353"
- "black hole entries" that get created when iptables/nftables _drops_ a packet, which record things like "packets from client 10.0.1.8:12345 to service 172.30.0.10:53 get dropped"

The DNAT entries cause connections to keep getting redirected to the same endpoint, which is what you want when the endpoint still exists, but is a problem when the endpoint goes away, so we manually delete the entries in that case. This is what the old `DeletedUDPEndpoints` code was for, and your new code deals with this.

The black hole entries cause connections to keep getting dropped, which is what you want when the service still has no endpoints, but is a problem once the service becomes able to accept connections, so we manually delete the entries in that case. This is what the old `NewlyActiveUDPServices` code was for, and your new code _doesn't_ deal with this.

Reply:
Also note that the old `NewlyActiveUDPServices` code deleted _all_ conntrack entries (both DNAT and black hole) for the relevant IPs, but that's only safe to do if you actually know that the service _just_ switched from 0 endpoints to non-0 endpoints. In the reconciler, if you only know that the service _currently_ has non-0 endpoints, but you don't know how many endpoints it had before, you need to be more careful that you only delete black-hole entries for the service, not DNAT entries.

Reply:
I tried to gather more information on this.

`nf_conntrack_in()` is registered with PREROUTING and OUTPUT hooks with priority -200. This puts the new connection in the `unconfirmed` table.
`nf_conntrack_confirm()` is registered with INPUT and POSTROUTING hooks with priority MAX. This moves the connection from the `unconfirmed` to `conntrack` table (the main table that we reconcile).

When a service has no endpoints we reject the packets service with (icmp-port-unreachable) and thus `nf_conntrack_confirm()` never gets executed for the packet and the entry from the `unconfirmed` table is deleted together with the packet reject. In some cases we drop rather than reject, both DROP and REJECT will prevent execution of `nf_conntrack_confirm()` and the packet will never be tracked by conntrack.

I created a UDP service without any endpoints and started pushing traffic to the ClusterIP, tried watching and listing the main conntrack table (using user-space binary) but couldn't find any event/flow. Tried to find the entry in the `unconfirmed` table but watching `unconfirmed` table is not allowed, and listing won't show the entry as the deletion from the "unconfirmed" table happens instantly with the packet reject.


If I remove the reject rule for services with no endpoint, then I see a conntrack entry in the main table
```
udp      17 29 src=10.244.2.3 dst=10.96.227.14 sport=58730 dport=3000 [UNREPLIED] src=10.96.227.14 dst=172.18.0.3 sport=3000 dport=58730 mark=0 use=1
```
Since no DNAT is performed `--reply-src` is the ClusterIP. On adding a new endpoint for the service, the reconciler clears the above flow entry as the logic is to clear all the flow entries that have `--orig-dest` as ClusterIP and `--reply-src` doesn't belong to any of the serving endpoint IP.

Immediately after the endpoint was in the serving state, I was able to see the following DNATed entry, and was able to communicate with the UDP service using the same client.
```
udp      17 27 src=10.244.2.3 dst=10.96.227.14 sport=58730 dport=3000 src=10.244.1.3 dst=10.244.2.3 sport=5000 dport=58730 mark=0 use=1
```

Reply:
> When a service has no endpoints we reject the packets service with (icmp-port-unreachable)

only for nftables

> Previously, before netlink/reconciler, we were doing cleanup using `--dst-nat` instead of using `--reply-src`, which I think clears only the DNATed destinations, maybe that is why cleanup for `NewlyActiveServices` was required. But with the DROP/REJECT rule, we won't need that cleanup as the conntrack entries won't be created.

Wait, I'm confused; I thought you said if you remove the reject rule, you get the UNREPLIED entry, but that it's OK because the new code removes that entry too, so it _will_ work even without the DROP/REJECT rule?

Reply:
> When a service has no endpoints we reject the packets service with (icmp-port-unreachable)

only for nftables

We add an explicit REJECT rule in iptales too. 
https://github.com/kubernetes/kubernetes/blob/948afe5ca072329a73c8e79ed5938717a5cb3d21/pkg/proxy/iptables/proxier.go#L1060-L1090

Saw this rule in iptables dump
```
-A KUBE-SERVICES -d 10.96.140.51/32 -p udp -m comment --comment "default/
udp has no endpoints" -m udp --dport 3000 -j REJECT --reject-with icmp-port-unreachable
```


IPVS there is no explicit iptable rule, but I observed ICMP `icmp-port-unreachable` in tcpdump. Couldn't see any conntrack entry either.

Reply:
> Wait, I'm confused; I thought you said if you remove the reject rule, you get the UNREPLIED entry, but that it's OK because the new code removes that entry too, so it will work even without the DROP/REJECT rule?

Yes, it works without without the DROP/REJECT rule.

Reply:
IPVS also sends `ICMP_PORT_UNREACH` when a virtual server has no real servers.
https://github.com/torvalds/linux/blob/v6.11/net/netfilter/ipvs/ip_vs_core.c#L654-L671a

All the linux proxiers triggers ICMP REJECT or packet DROP(special case for non-local-endpoints when policies are set) when there are no endpoints for a VIP.

Reply:
> "black hole entries" that get created when iptables/nftables drops a packet, which record things like "packets from client 10.0.1.8:12345 to service 172.30.0.10:53 get dropped"

Terminology is misleading, let's call black hole entries to the conntrack stale entries that drop traffic that should not drop. AFAIK, and I also tested to verify it, there is no such conntrack entries for drop/reject rules , it is the netfilter hook  that drops or reject the packets, but nothing is populated in the conntrack table about these connections.

Reply:
> "DNAT entries" that get created when iptables/nftables DNATs a packet, which record things like "packets from client 10.0.1.8:12345 to service 172.30.0.10:53 get NATted to endpoint 10.128.0.5:5353"

These are also back-holing entries, right? 
When the service changes the endpoint and we don't clear the entries DNATed to the previous endpoint traffic will be black-holed as the endpointIP no longer exists.

Reply:
> We add an explicit REJECT rule in iptables too.

oh, duh, I was confusing the no-endpoints reject with the invalid-port reject (#122692)

Reply:
OK, the code that eventually become `NewlyActiveUDPServices` was added in #48524 to fix #48370, which claims

> Change service to have zero endpoints by killing the pod, we get the following conntrack entry:
>
>     udp      17 29 src=10.12.0.14 dst=10.15.240.10 sport=43145 dport=53 [UNREPLIED] src=10.15.240.10 dst=10.12.0.14 sport=53 dport=43145 mark=0 use=1

and examining the tree at that point shows that [we _did_ have the REJECT rule for services with no endpoints](https://github.com/kubernetes/kubernetes/blob/25ac521f88da270f8ac2443f8974963b45b471bc/pkg/proxy/iptables/proxier.go#L1402-L1413). (It was added more than a year earlier,  #19576.)

So I'm not sure what has changed that you are only seeing an `UNREPLIED` entry when there is no reject rule.

Reply:
I double-checked, and this time ran a kind cluster with kindest/node:v1.31.0 image.
Created a UDP service with no endpoints, started pushing packets and didn't see any UNREPLIED entry.

Removed kube-proxy daemon set from the cluster and manually deleted the REJECT rule.
```
root@kind-worker:/# iptables -D KUBE-SERVICES -d 10.96.90.248/32 -p udp -m comment --comment "default/udp has no endpoints" -m udp --dport 3000 -j REJECT --reject-with icmp-port-unreachable
```
and saw the UNREPLIED entry
```
root@kind-worker:/# conntrack -Ln -p udp | grep -v port=53
conntrack v1.4.7 (conntrack-tools): 1 flow entries have been shown.
udp      17 29 src=10.244.3.2 dst=10.96.90.248 sport=36000 dport=3000 [UNREPLIED] src=10.96.90.248 dst=172.18.0.3 sport=3000 dport=36000 mark=0 use=1
```

Maybe something changed in conntrack module in the recent kernel versions 🤔 ?